feat: remirror for note editing

mtfoley opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/open-sauced/blob/HEAD/CONTRIBUTING.md#create-a-pull-request.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/open-sauced/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [x] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

Congrats to @ifiokjr on release of v1.0.0 of remirror! 
I wanted to get started on a PR draft for the remirror note editing feature, since @0-vortex helped tremendously in cleaning up npm dependencies to free up this feature and others.

Todo:
- [x] Support and render some subset of markdown in editor
- [x] Manage state and initial content correctly, I'm still trying to get my head around the docs on this [https://remirror.io/docs/react/controlled](https://remirror.io/docs/react/controlled)
- [x] Implement social editor features. 
- [ ] ~~Add to storybook~~ NoteForm isn't currently in StoryBook (makes sense for open form control?)
- [x] Add/update docs
- [ ] ~~Add test similar to what's in [https://github.com/remirror/remirror/blob/main/packages/remirror__extension-markdown/__tests__/markdown-extension.spec.ts](https://github.com/remirror/remirror/blob/main/packages/remirror__extension-markdown/__tests__/markdown-extension.spec.ts). I'm skipping the existing NoteForm test on this first commit.~~ Sticking with "no violations on render" test?

This PR closes #769

## Related Tickets & Documents
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

## Mobile & Desktop Screenshots/Recordings
![Screen Shot 2021-07-20 at 10 50 56 PM](https://user-images.githubusercontent.com/3792749/126423023-beb86b4b-7cb4-4ef9-b40a-42938a244911.png)

<!-- Visual changes require screenshots -->


## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 docs
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
![turtle going up a slide](https://media.giphy.com/media/w87yLYL7lwDWE/giphy.gif)

<!-- note: PRs with deletes sections will be marked invalid -->

mtfoley commented:
@bdougie @0-vortex @filiptronicek looks like Netlify deploy is choking on my CSS import in `src/components/NoteFormSocial.js`. Works ok for Webpack when running `npm start` locally, so I don't know what I'll need to do to make this work. I'd be referring to the remirror docs to get this working: [https://remirror.io/docs/installation#styling-the-editor](https://remirror.io/docs/installation#styling-the-editor)

0-vortex commented:
I'll have a swing at it in the morning 🤤

0-vortex commented:
> Installing remirror gives you access to all the extensions and core packages that are actively developed in the remirror repository.

Installing packages from both sources makes it very difficult to figure out exactly what the tree should look like - suggest installing just the required packages if you want a readable tree :D

ifiokjr commented:
@mtfoley I'll be able to give you a hand on this in a few hours time 😊

mtfoley commented:
Three steps forward, two steps back - deploy preview works on Netlify now. I added the `MentionExtension` and the `EmojiExtension`.  The MentionExtension doesn't seem to be doing anything, but I'm not sure if I need to feed it a set of valid names for mentioning. The EmojiExtension makes the app crash when running locally and I trigger it with by typing the activation character `:`.

@ifiokjr any pointers you'd like to give, would love it!

mtfoley commented:
FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
![Screen Shot 2021-08-01 at 10 08 28 PM](https://user-images.githubusercontent.com/3792749/127795344-dd0a30a1-b744-44ec-b863-ae6b0dc45de7.png)

0-vortex commented:
> FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
> ![Screen Shot 2021-08-01 at 10 08 28 PM](https://user-images.githubusercontent.com/3792749/127795344-dd0a30a1-b744-44ec-b863-ae6b0dc45de7.png)

gonna pick this up after work and let you know !

mtfoley commented:
I hit up the remirror discord so we will see if anything comes back.

0-vortex commented:
> FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
> ![Screen Shot 2021-08-01 at 10 08 28 PM](https://user-images.githubusercontent.com/3792749/127795344-dd0a30a1-b744-44ec-b863-ae6b0dc45de7.png)

this is because of https://github.com/remirror/remirror/blob/5764fde3bc4643cb8c6de57ce4f3d2baa3654361/packages/remirror__extension-emoji/src/emoji-extension.ts#L37-L47 the emoji selection character is the end selection handler for remirror - configuring this plugin to use a different shortcut will remove this error but I have not produced a full working build yet

mtfoley commented:
Got this back on the remirror discord:
> Could you try this?

```js
import data from 'svgmoji/emoji.json';

new EmojiExtension({ data, moji: 'noto' })
```

> I believe the README of `extension-emoji` has some mistakes. Will fix.

mtfoley commented:
Ok I kind of got things to work with a snippet like that. However the suggestion box doesn’t come up but it will render an emoji you type. Trouble is, it ends up being added as an svg image rather than as the emoji itself so the markdown has these huge images interspersed with the text. The other side effect of this is that when you open the editor a second time to edit the notes the emojis are effectively lost because it seems I don’t have the editor ready to support those. I’ll push the changes up later today so y’all can see what im talking about.

mtfoley commented:
Emojis work much better with plainText:true option. Still not seeing suggestions popup yet.

Other kink to mess with is whether editor will work correctly in dark mode. if you type ## then hit space it starts using heading 2 but in black text on dark background. Need to see if there's some step required to make markdown extension aware of colors in use... maybe?

mtfoley commented:
Got emoji suggestion popup working, and changed gears on styling method in the process (using `AllStyledComponent` instead of CSS import).  I found this closed issue in the remirror repo and based on the fact that the person wasn't told they needed to make any changes, I assumed they were pretty close: https://github.com/remirror/remirror/issues/968

Emoji suggestion popup has a transparent background which I'm not sure is right. It's also staying inside the editor's outline, so I'm not sure that's right either. 
![Screen Shot 2021-08-07 at 3 16 00 PM](https://user-images.githubusercontent.com/3792749/128611575-07d9baa9-35ac-449f-92af-07b9987384e7.png)


I did find that remirror styles headings with a CSS color variable, so I may be able to use some CSS rules to account for this.

I think another change that I'm going to need is to pull the "Edit Notes" and "Cancel" buttons inside the Remirror component so that I can mess with state more easily 🤞 

This thing is turning into spaghetti, so I'll probably break it up into a component once something works well.

mtfoley commented:
Progress so far:
- added CSS rules to handle menu background color and making editor bigger so more of the menu would fit.  
- preloading the SVG image from with `display:none` so that the emoji menu is snappier when it's first used. 
- added the code extension too in order to allow inline code blocks. I don't think it does multiline or language specific blocks.

mtfoley commented:
@0-vortex I applied those fixes you put up the other day, and merged latest upstream 🙏

If anyone can take a look at my state mgmt I'd appreciate it.  When you click cancel while editing, the notes go blank.  Apparently I'm not setting "previouslySavedValue" correctly.

0-vortex commented:
> @0-vortex I applied those fixes you put up the other day, and merged latest upstream 🙏
> 
> If anyone can take a look at my state mgmt I'd appreciate it. When you click cancel while editing, the notes go blank. Apparently I'm not setting "previouslySavedValue" correctly.

Can take a look after work, not thrilled about state management so maybe link all the issues and we finish it today? Maybe get @bdougie to look at this on stream?

mtfoley commented:
@0-vortex Finally peer pressured me into taking this thing out of draft status. Last bit of context:
- I was trying out tests last night. Two snags for replicating existing tests, and I think we'd need to add another dependency to replicate test as-is and I'm not sure how it would actually work. At the moment it passes the "no violations" test.
Still need to add to storybook.

0-vortex commented:
### Tooling wise, how to make your PR the best PR possible 🔥 

#### Step 1: restore the proper lockfiles

```shell
git restore --source main -- npm-shrinkwrap.json package.json
```


#### Step 2: remirror perfect life

Then, remember all the support you got from `@remirror` discord and start to verify their `package.json` and go full minimalist, installing only the absolutely required packages.

```shell
npm i -S @remirror/react @remirror/styles @remirror/extension-bold @remirror/extension-code @remirror/extension-heading @remirror/extension-italic @remirror/extension-list @remirror/extension-markdown @remirror/extension-tables
```


#### Step 3: remove the `remirror/extensions` import

```javascript
import {CodeExtension} from "@remirror/extension-code";
import {BulletListExtension} from "@remirror/extension-list";
import {HeadingExtension} from "@remirror/extension-heading";
import {BoldExtension} from "@remirror/extension-bold";
import {ItalicExtension} from "@remirror/extension-italic";
import {TableExtension} from "@remirror/extension-react-tables";
import {MarkdownExtension} from "@remirror/extension-markdown";
import {EmojiExtension} from "@remirror/extension-emoji";
```

0-vortex commented:
LGTM!

mtfoley commented:
in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛  Pushed changes and removed old files.

Current form isn't in Storybook. Keeping no violations test pending review.

0-vortex commented:
> in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> 
> Current form isn't in Storybook. Keeping no violations test pending review.

What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱

mtfoley commented:
> > in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> > Current form isn't in Storybook. Keeping no violations test pending review.
> 
> What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱

Says it in the docs [https://git-scm.com/docs/git-restore](https://git-scm.com/docs/git-restore)... which I only read bc my version of git apparently wasn't new enough so I ended up doing `wget -O package.json https://raw.githubusercontent.com/open-sauced/open-sauced/main/package.json` 🥇

mtfoley commented:
BTW, still gotta update README.md at least...

0-vortex commented:
> > > in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> > > Current form isn't in Storybook. Keeping no violations test pending review.
> > 
> > 
> > What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱
> 
> Says it in the docs https://git-scm.com/docs/git-restore... which I only read bc my version of git apparently wasn't new enough so I ended up doing `wget -O package.json https://raw.githubusercontent.com/open-sauced/open-sauced/main/package.json` 🥇

Oh snap I apparently don't have experimental in my vocabulary, that's a good first-time issue for updating the pull request docs. Or you could do it since you discovered it! 👯

mtfoley commented on src/components/NoteFormSocial.js:
```diff
@@ -0,0 +1,150 @@
+  // Access the commands and the activity status of the editor.
+  const commands = useCommands();
+  const startNewNote = () => {
+    commands.insertMarkdown("### Notes for Repo\n- fork the repo\n\n- follow everybody on twitter\n-troll, troll, troll<Cursor>");
```
@bdougie this is a bit of an easter egg. anyway, I'm undecided about whether to keep this button in place (@0-vortex had suggested changing the label on it), as opposed to just pre-loading some template markdown when someone starts editing a blank note.

0-vortex replied:
freedom!

bdougie replied:
Let's name the button, "basic template"

mtfoley replied:
👍

bdougie commented on src/index.css:
```diff
@@ -117,4 +117,20 @@ body.dark .react-loading-skeleton{
 }
-
+.remirror-editor {
+  min-height: 250px !important;
```
Why the important here. Is this because remirror does not play nicE?

mtfoley replied:
Editor stayed small and emoji list stayed too small to be usable. you would only see like 3 options

bdougie reviewed (commented):
Just one question then I can merge

bdougie commented on src/index.css:
```diff
@@ -117,4 +117,20 @@ body.dark .react-loading-skeleton{
   );
 }
-
+.remirror-editor {
```
```suggestion
// important is to manage the emoji sizing, so more appear while searching. 
.remirror-editor {
```

bdougie reviewed (approved):
Looking good, we will need to open issues on the UI and CSS.
